File: services/api/src/api/routers/usage.py
"""Usage summary (web console Phase 2): per-key request counts from key_usage.

One GROUP BY over key_usage joined to api_keys, bounded by a period cutoff.
At our scale (~few hundred rows/day) a live aggregate is fine (PRD §11).
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Literal

from core.db.models import ApiKey, KeyUsage
from fastapi import APIRouter, Depends
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from api.deps import get_session, require_scope

router = APIRouter(prefix="/v1/usage", tags=["usage"])

Period = Literal["today", "24h", "7d", "30d"]
_PERIOD_HOURS = {"24h": 24, "7d": 24 * 7, "30d": 24 * 30}


def _cutoff(period: Period, now: datetime) -> datetime:
    if period == "today":
        return now.replace(hour=0, minute=0, second=0, microsecond=0)
    return now - timedelta(hours=_PERIOD_HOURS[period])


@router.get("/summary")
def usage_summary(
    period: Period = "24h",
    key=Depends(require_scope("admin")),
    session: Session = Depends(get_session),
):
    """{period, total, by_key: [{key_id, name, calls, last_used_at}]}, calls desc."""
    start = _cutoff(period, datetime.now(UTC))
    rows = session.execute(
        select(
            ApiKey.id,
            ApiKey.name,
            func.count(KeyUsage.id).label("calls"),
            func.max(KeyUsage.created_at).label("last"),
        )
        .join(KeyUsage, KeyUsage.api_key_id == ApiKey.id)
        .where(KeyUsage.created_at >= start)
        .group_by(ApiKey.id, ApiKey.name)
        .order_by(func.count(KeyUsage.id).desc())
    ).all()
    return {
        "period": period,
        "total": int(sum(r[2] for r in rows)),
        "by_key": [
            {
                "key_id": str(r[0]),
                "name": r[1],
                "calls": int(r[2]),
                "last_used_at": r[3],
            }
            for r in rows
        ],
    }

File: internal/api/usage.go
package api

import (
	"net/http"
	"time"
)

// usageRow is one by_key entry (usage.py shape).
type usageRow struct {
	KeyID      string     `json:"key_id"`
	Name       *string    `json:"name"`
	Calls      int        `json:"calls"`
	LastUsedAt *time.Time `json:"last_used_at"`
}

// usageSummary is the /v1/usage/summary payload: {period, total, by_key},
// calls desc.
type usageSummary struct {
	Period string     `json:"period"`
	Total  int        `json:"total"`
	ByKey  []usageRow `json:"by_key"`
}

// periodCutoff implements usage.py's cutoff semantics:
// today → local midnight; 24h/7d/30d → now minus N hours.
func periodCutoff(period string, now time.Time) (time.Time, bool) {
	switch period {
	case "today":
		y, m, d := now.Date()
		return time.Date(y, m, d, 0, 0, 0, 0, now.Location()), true
	case "24h":
		return now.Add(-24 * time.Hour), true
	case "7d":
		return now.Add(-24 * 7 * time.Hour), true
	case "30d":
		return now.Add(-24 * 30 * time.Hour), true
	}
	return time.Time{}, false
}

func (s *Server) handleUsageSummary(w http.ResponseWriter, r *http.Request) {
	period := r.URL.Query().Get("period")
	if period == "" {
		period = "24h"
	}
	cutoff, ok := periodCutoff(period, time.Now())
	if !ok {
		writeDetail(w, http.StatusUnprocessableEntity,
			"period must be one of today|24h|7d|30d")
		return
	}
	rows, err := s.deps.DB.UsageSummary(r.Context(), cutoff)
	if err != nil {
		writeDetail(w, http.StatusInternalServerError, err.Error())
		return
	}
	out := usageSummary{Period: period, ByKey: []usageRow{}}
	for _, row := range rows {
		out.Total += row.Calls
		out.ByKey = append(out.ByKey, usageRow{
			KeyID: row.KeyID, Name: row.Name, Calls: row.Calls, LastUsedAt: row.LastUsedAt,
		})
	}
	WriteJSON(w, http.StatusOK, out)
}
